internal/helm/templatetypes/types.go

```diff
@@ -30,6 +30,73 @@ type DurationOrTemplate struct {
 	ValueTemplate string         `yaml:",omitempty"`
 }
 
+// StringTemplate contains a string that can be a template value
+type StringTemplate struct {
+	Value string
+}
+
+// StringTemplateList is a list of StringTemplate
+type StringTemplateList []StringTemplate
+
+// ShellCommandTemplate is a shell command parsed as a list or string
+type ShellCommandTemplate []StringTemplate
+
+// HostsListTemplate is a list to hosts parsed as a list or map
+type HostsListTemplate []StringTemplate
+
+// LabelsTemplate is a mapping type for labels
+type LabelsTemplate map[StringTemplate]StringTemplate
+
+// MappingWithEqualsTemplate is a mapping type parsed as a list or map
+type MappingWithEqualsTemplate map[StringTemplate]*StringTemplate
+
+// ProcessTemplate can be overridden to process template values when marshaling
+var ProcessTemplate = func(s string) (string, error) {
```

Comment:
I am really not fan of that variable overidden above when running it… This package is only used for the helm stuff, it's in `internal`, so I don't see any value of allowing customization, or I am missing something…

Reply:
there could be other uses in the future for other source-to-source transfrom use cases, and despite being in a helm subdir templatetypes has nothing helm-specific in it, so I think it's cleaner that way.

Reply:
Honestly I don't… as said "there *could* be uses …", but as of today there is not… this is call [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

> Always implement things when you actually need them, never when you just foresee that you need them.

Reply:
Looking at your front wheel is the best way to fall when you're cycling. Same thing for computer science :p 
More seriously it's mainly about clean separation of concern. I don't want the helm template handler code to sit in package templatetypes. It has nothing to do there.

Reply:
ISTM that the _real_ issue here is that the YAML library has no way to carry any state or context from the top-level `Marshal` call (or `Encoder` object) to the underlying types implementing the custom `Marshaller` interface. Which means that if you want your types (which are generic) to have a dynamic marshalling behaviour based somehow on context you are forced to have some sort of global state, either a global `var` (potentially racy) or a (set of) global `func` (which leaks some specific variant(s) of the contextual requirement(s) into the more generic `Marshaller` implementation for the generic underlying type).

Neither of those situations are especially palatable but without patching `go-yaml` (e.g. to pass a `userdata interface{}` to those callbacks) we are basically stuck with those choices AFAICT.

Given that `var` and `func` here are basically both equally bad and this PR is already written using one of them it doesn't seem like there is much reason to change how things are done.

(I suppose one might want to argue for a lock around the use of the global var or something, that at least mitigates some of the badness in a very small way)

(obv the above all applies to unmarshalling too)